integration/v3_lock_test.go

```diff
@@ -101,6 +102,107 @@ func TestMutexSessionRelock(t *testing.T) {
 	}
 }
 
+// TestMutexWaitsOnCurrentHolder ensures
+// all lock requests older than the owner's lock request are gone.
+// e.g a series of Waiters A <- B <- C where A holds the lock
+// and B waits before C. The test case ensures that
+// the failure of B won't result C to obtain the lock, and C only
+// obtains lock when lock requests A and B are gone through releasing lock or expired lease.
+func TestMutexWaitsOnCurrentHolder(t *testing.T) {
+	clus := NewClusterV3(t, &ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+
+	cctx, ccancel := context.WithCancel(context.Background())
+
+	cli := clus.Client(0)
+
+	s1, err := concurrency.NewSession(cli)
+	if err != nil {
+		t.Error(err)
+	}
+	firstMutex := concurrency.NewMutex(s1, "test-mutex")
+	if err := firstMutex.Lock(cctx); err != nil {
+		t.Fatal(err)
+	}
+
+	victimSession, err := concurrency.NewSession(cli)
+	if err != nil {
+		t.Error(err)
+	}
+	victimMutex := concurrency.NewMutex(victimSession, "test-mutex")
+	vmDonec := make(chan struct{})
+	go func() {
+		defer close(vmDonec)
+		victimMutex.Lock(cctx)
+	}()
+
+	// ensure victimMutex waits before ownerMutex
+	wch := cli.Watch(cctx, "test-mutex", clientv3.WithPrefix(), clientv3.WithRev(1))
+	putCounts := 0
+	for putCounts < 2 {
+		select {
+		case wrp := <-wch:
+			putCounts += len(wrp.Events)
+		case <-time.After(time.Second):
+			t.Fatal("failed to receive watch response")
+		}
+	}
+	if putCounts != 2 {
+		t.Fatalf("expect 2 put events, but got %v", putCounts)
+	}
+
+	ownerSession, err := concurrency.NewSession(cli)
+	if err != nil {
+		t.Error(err)
+	}
+	ownerMutex := concurrency.NewMutex(ownerSession, "test-mutex")
+	ownerDonec := make(chan struct{})
+	go func() {
+		defer close(ownerDonec)
+		ownerMutex.Lock(cctx)
+	}()
+
+	// close victimSession to simulate a node failure
+	// which will result deletion of victimMutex waiter.
+	victimSession.Close()
+
+	// ensures the deletion of victimMutex waiter from server side.
+	wch = cli.Watch(cctx, "test-mutex", clientv3.WithPrefix(), clientv3.WithRev(4))
+	select {
+	case wrp := <-wch:
+		e := wrp.Events[0]
+		if e.Type != mvccpb.DELETE {
+			t.Fatalf("expect a delete event on prefix test-mutex, but got event type %v", e.Type)
+		}
+	case <-time.After(time.Second):
+		t.Fatal("failed to receive a watch response")
+	}
+
+	select {
+	case <-ownerDonec:
+		t.Fatal("Owner obtained lock without previous lock requests being gone")
+	default:
+	}
+
+	if err := firstMutex.Unlock(cctx); err != nil {
+		t.Fatal(err)
+	}
+
+	select {
+	case <-ownerDonec:
+	case <-time.After(time.Second):
+		t.Fatal("owner failed to obtain lock")
+	}
+
+	// release resource of victimMutex
+	ccancel()
```

Comment:
I don't think this is necessary. The victim mutex is waiting on the first owner. Once the first owner is unlocked, the victim mutex will think it obtained the lock and proceed to exit its goroutine in the test.